File: rotate.py
#!/usr/bin/env python3
import sys

def rotate_cell_around_centre_general(cell,centre):

    rotated = [centre[0]+centre[1]-cell[1],-centre[0]+2*centre[1]+cell[0]-cell[1]]
    return rotated

def rotate_cell_CEE(cell):

    rotated = [-cell[1],cell[0]-cell[1]]
    return rotated

def rotate_cell_CEH_Odd(cell):

    rotated = [-1-cell[1],-1+cell[0]-cell[1]]
    return rotated

def rotate_cell_CEH_Even(cell):

    rotated = [1-cell[1],1+cell[0]-cell[1]]
    return rotated

def rotate(u,v,layer,sector):
    uv = [u,v]

    if ( sector == 0 ):
        return uv
    
    if (layer > 28):
        if (layer % 2) == 0:

            if ( sector == 1 ):
                return rotate_cell_CEH_Even(uv)
            elif ( sector == 2 ):
                return rotate_cell_CEH_Even(rotate_cell_CEH_Even(uv))

        else:

            if ( sector == 1 ):
                return rotate_cell_CEH_Odd(uv)
            elif ( sector == 2 ):
                return rotate_cell_CEH_Odd(rotate_cell_CEH_Odd(uv))

    else:
        if ( sector == 1 ):
            return rotate_cell_CEE(uv)
        elif ( sector == 2 ):
            return rotate_cell_CEE(rotate_cell_CEE(uv))

            
def rotate_to_sector_0(u,v,layer):

    sector = 0
    uv = [u,v]

    if (layer > 28):

        if (layer % 2) == 0: #Even CEH Layers

            if ( u > 0 and v > 0 ): #sector 0
                return uv,sector
            elif ( u <= 0 and v > u ): #sector 1
                sector = 1
                return rotate_cell_CEH_Even(rotate_cell_CEH_Even(uv)),sector
            else: #sector 2
                sector = 2
                return rotate_cell_CEH_Even(uv),sector
            
        else: #Odd CEH Layers
                
            if ( u >= 0 and v >= 0 ): #sector 0
                return uv,sector
            elif ( u < 0 and v >= u ): #sector 1
                sector = 1
                return rotate_cell_CEH_Odd(rotate_cell_CEH_Odd(uv)),sector
            else: #sector 2
                sector = 2
                return rotate_cell_CEH_Odd(uv),sector

    else: #CEE

        if ( u > 0 and v >= 0 ): #sector 0
            return uv,sector
        elif ( u <= 0 and v > u ): #sector 1
            sector = 1
            return rotate_cell_CEE(rotate_cell_CEE(uv)),sector
        else: #sector 2
            sector = 2
            return rotate_cell_CEE(uv),sector

#Main function for standalone use

# def main():
    
#     u = int(sys.argv[1])
#     v = int(sys.argv[2])
#     layer = int(sys.argv[3])

#     uv = rotate_to_sector_0(u,v,layer)
#     print ( "python: " +  str(uv[0]) + "," + str(uv[1]) )


# main()

File: rotate.cxx
#include <iostream>


void uvMapping(unsigned layer, std::pair<int,int> &uv) {
 unsigned sector(0);
 int offset;

 if(layer<=28) { // CE-E    
   if(uv.first>0 && uv.second>=0) return;

   offset=0;
   if(uv.first>=uv.second && uv.second<0) sector=2;
   else sector=1;

 } else if((layer%2)==1) { // CE-H Odd
   if(uv.first>=0 && uv.second>=0) return;

   offset=-1;    
   if(uv.first>uv.second && uv.second<0) sector=2;
   else sector=1;

 } else { // CE-H Even
   if(uv.first>=1 && uv.second>=1) return;

   offset=1;
   if(uv.first>=uv.second && uv.second<1) sector=2;
   else sector=1;
 }

 int up,vp;

 if(sector==1) {
   up=uv.second-uv.first;
   vp=-uv.first+offset;    

 } else {
   up=-uv.second+offset;
   vp=uv.first-uv.second+offset;
 }

 uv.first=up;
 uv.second=vp;
 return;
}

int main(int argc, char **argv){

  int u = atoi(argv[1]);
  int v = atoi(argv[2]);
  unsigned layer = atoi(argv[3]);
  std::pair<int,int> uv = std::make_pair(u,v);

  uvMapping(layer, uv);

  std::cout << "c++: " << uv.first << "," << uv.second << std::endl;
  
  return 0;
}
